volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+    ) -> Generator[Tuple[str, int], None, None]:
+        """
+        Enumerates the symbols specified as wanted by the calling plugin
+        """
+        wanted_symbols = wanted_modules[mod_name]
+
+        if (
+            PESymbols.wanted_names not in wanted_symbols
+            and PESymbols.wanted_addresses not in wanted_symbols
+        ):
+            vollog.warning(
+                f"Invalid `wanted_symbols` sent to `find_symbols` for module {mod_name}. addresses and names keys both misssing."
+            )
+            return
+
+        symbol_keys = [
+            (PESymbols.wanted_names, "get_address_for_name"),
+            (PESymbols.wanted_addresses, "get_name_for_address"),
+        ]
+
+        for symbol_key, symbol_getter in symbol_keys:
+            # address or name
+            if symbol_key in wanted_symbols:
+                # walk each wanted address or name
+                for wanted_value in wanted_symbols[symbol_key]:
+                    symbol_value = symbol_resolver.__getattribute__(symbol_getter)(
+                        wanted_value
+                    )
+                    if symbol_value:
+                        # yield out symbol name, symbol address
+                        if symbol_key == PESymbols.wanted_names:
+                            yield wanted_value, symbol_value  # type: ignore
+                        else:
+                            yield symbol_value, wanted_value  # type: ignore
+
+                        index = wanted_modules[mod_name][symbol_key].index(wanted_value)  # type: ignore
+
+                        del wanted_modules[mod_name][symbol_key][index]
+
+                # if all names or addresses from a module are found, delete the key
+                if not wanted_modules[mod_name][symbol_key]:
+                    del wanted_modules[mod_name][symbol_key]
+                    break
+
+    @staticmethod
+    def _resolve_symbols_through_methods(
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+        module_instances: List[Tuple[str, int, int]],
+        wanted_modules: PESymbolFinder.cached_value_dict,
+        mod_name: str,
+    ) -> Generator[Tuple[str, int], None, None]:
+        """
+        Attempts to resolve every wanted symbol in `mod_name`
+        Every layer is enumerated for maximum chance of recovery
+        """
+        symbol_resolving_methods = [
+            PESymbols._find_symbols_through_pdb,
+            PESymbols._find_symbols_through_exports,
+        ]
+
+        for method in symbol_resolving_methods:
+            for symbol_resolver in method(
+                context, config_path, module_instances, mod_name
+            ):
+                vollog.debug(f"Have resolver for method {method}")
+                yield from PESymbols._get_symbol_value(
+                    wanted_modules, mod_name, symbol_resolver
+                )
+
+                if not wanted_modules[mod_name]:
```

Comment:
So, we yielded, but if it's still present, then we break?  So basically, if it wasn't the first hit (because we're still yielding and it may get removed from wanted_modules on a second run)?  So then we swap to a different resolver method, unless we didn't find it and then we've broken again.  So this gives up as soon as the first one says it can't do it?  So it only traverses the loops until the first failure?  The docstring definitely needs to talk through what this is doing based on the inputs and outputs.

Reply:
This is saying to break if we the module is now removed, meaning all the wanted symbols were found inside of it. I can add more comments to that effect, but it makes to where the first context (process in userland) that holds the needed data (PDB sig or export table) can resolve all the wanted symbols then the processing loop stops.

Reply:
Ok, this looked like it was doing it just for `wanted_modules[mod_name]` rather than all the wanted symbols, but as I said, the variables flying around confused me and it was really hard to follow what lived where.  Would be really good to have an ascii diagram of something showing the lists of tables of whatever of how stuff is being stored...  https://asciiflow.com/#/ or something just so people can get an idea of what lives in a fully populated data structure...

Reply:
In comments, I added plugins to refer to for complete examples. I am pretty sure if you looked at the plugins first (debug regsiters and un hooked system calls) that you would have got through the API understanding alot easier :(